katdal/sensordata.py

```diff
@@ -249,37 +388,48 @@ def dummy_sensor_data(name, value=None, dtype=np.float64, timestamp=0.0):
             value = ''
         elif np.issubdtype(dtype, np.bool):
             value = False
+        elif np.issubdtype(dtype, np.object):
+            value = None
     else:
         dtype = np.array(value).dtype
-    data = np.rec.fromarrays([[timestamp], [value], ['nominal']],
-                             dtype=[('timestamp', np.float64), ('value', dtype), ('status', '|S7')])
-    return SensorData(data, name)
+    data = np.array(zip([timestamp], [value]),
+                    dtype=[('timestamp', np.float64), ('value', dtype)])
+    return RecordSensorData(data, name)
 
 
-def remove_duplicates(sensor):
-    """Remove duplicate timestamp values from sensor data.
+def remove_duplicates_and_failures(sensor):
+    """Remove duplicate timestamp values and failures from sensor data.
 
     This sorts the 'timestamp' field of the sensor record array and removes any
     duplicate values, updating the corresponding 'value' and 'status' fields as
     well. If more than one timestamp have the same value, the value and status
     of the last of these timestamps are selected. If the values differ for the
     same timestamp, a warning is logged (and the last one is still picked).
 
+    In addition, if there is a 'status' field, get rid of data with a 'failure'
+    status which indicates that the sensor could not be read and the
+    corresponding value will therefore be meaningless. Afterwards, remove the
+    'status' field from the data as this is the only place it plays a role.
```

Comment:
There are multiple statuses that mean the value is meaningless (see page 20 of the [katcp spec](https://pythonhosted.org/katcp/_downloads/NRF-KAT7-6.0-IFCE-002-Rev5.pdf)). It's better to assume that anything that isn't one of `nominal`, `warn`, `error` has no meaningful value.